cpp-client/deephaven/dhclient/include/private/deephaven/client/impl/client_impl.h

```diff
@@ -27,14 +29,24 @@ class ClientImpl {
   ClientImpl(Private, std::shared_ptr<TableHandleManagerImpl> &&manager_impl);
   ~ClientImpl();
 
-  void Shutdown() {
-    managerImpl_->Shutdown();
-  }
+  void Shutdown();
 
   [[nodiscard]]
-  const std::shared_ptr<TableHandleManagerImpl> &ManagerImpl() const { return managerImpl_; }
+  const std::shared_ptr<TableHandleManagerImpl> &ManagerImpl() const { return manager_impl_; }
+
+  using OnCloseCbId = utility::OnCloseCbId;
+  using OnCloseCb = utility::OnCloseCb;
+
+  OnCloseCbId AddOnCloseCallback(OnCloseCb cb);
+  bool RemoveOnCloseCallback(OnCloseCbId cb_id);
 
 private:
-  std::shared_ptr<TableHandleManagerImpl> managerImpl_;
+  std::shared_ptr<TableHandleManagerImpl> manager_impl_;
+  mutable std::mutex on_close_mux_;
+  struct OnCloseCbContext {
+    std::uint32_t next_id_;
+    std::map<OnCloseCbId, OnCloseCb> map_;
+  };
+  std::unique_ptr<OnCloseCbContext> on_close_cb_ctx_;
```

Comment:
Why do we have a unique_ptr here rather than
```
OnCloseCbContext on_close_cb_ctx;
```

Or just promote the two fields into the enclosing class and get rid of the `struct OnCloseCbContext` altogether.

I don't see any advantage to the unique_ptr, and the disadvantage is extra code to do null tests. The fields (a uint32_t and a map) are pretty small so I don't think we should be worried about making ClientImpl slightly larger.